### ISSUE TITLE
Move HTTP internal only components to the Internal namespace

### ISSUE THREAD

**Issue by vhvb1989** [Azure/azure-sdk-for-cpp#1342]
There are some HTTP components which are not expected to be used by end-users but only for services.

- HttpPipeline
- HttpRequest
- HttpUrl

However, there is currently a dependency problem that need to be addressed.

The HttpTransport base class defines the `send` method which requires an HttpRequest:
```cpp
std::unique_ptr<RawResponse> Send(Context const& context, Request& request) override;
```

If the HttpTranport is moved to the internal namespace, anyone interested in implementing a new Transport Adapter would need to consume the Internal namespace.

If the HttpTransport remains public, it would need to consume and expose the Request class from the internal namespace.

@JeffreyRichter  @RickWinter 
Please take a look and comment about the options.

The last option would be to let all Http components as public (Like it currently is)

**Comment by ahsonkhan** [comment]
> If the HttpTranport is moved to the internal namespace, anyone interested in implementing a new Transport Adapter would need to consume the Internal namespace.

> If the HttpTransport remains public, it would need to consume and expose the Request class from the internal namespace.

Neither option seems ideal.

> The last option would be to let all Http components as public (Like it currently is)

I like this :)
I am not suggesting we do this, but what about keeping `Request` public but making `HttpPipeline` internal, as another option?

**Comment by vhvb1989** [comment]
> but what about keeping Request public but making HttpPipeline internal, 

The pipeline also uses the Request. It exposes a public `Send()`  method to start the pipeline

**Comment by ahsonkhan** [comment]
That's fine, right? An internal type can take a dependency on a public type. So, why would making `pipeline` internal be an issue (even if it exposes Request), when Request is already public.

**Comment by JeffreyRichter** [comment]
If customers create their own pipeline policies, then these need to be public for them

**Comment by vhvb1989** [comment]
> If customers create their own pipeline policies, then these need to be public for them

They would include policy.hpp header which defines an HttpPolicy to create a new policy.  However, customer can't create and start a pipeline directly. For that, they need to use an SDK client.
The SDK client should provide a way to add custom policies to it

**Comment by JeffreyRichter** [comment]
The customer's policy would be passed a Request. Doesn't this expose fields/methods that use HttpUrl and/or HttpRequest?

**Comment by RickWinter** [comment]
We should not place anything in the Internal namespace that a customer may use. 

@vhvb1989 Are there other questions on this topic or is this issue able to be closed?

**Comment by vhvb1989** [comment]
@RickWinter The http pipeline is not used by end users, only by service clients. This issue requires to move the pipeline class to be internal.

We also mentioned before that all policies should be internal as well, since end users are not expected to use them. Only the base class Policy would be public for end users to be able to create policies. Then they need to use an sdk client to add those policies

**Comment by vhvb1989** [comment]
Closed by #1642

**Comment by JeffreyRichter** [comment]
We use the internal namespace for things that we offer to other Azure service teams building client libraries whose code also resides in our repo. The internal abstractions are not well documented and are designed to support only the scenarios that Azure service teams require; they are not designed in a general-purpose way and we do not support them for general consumption. 

So, you can use our internal abstractions but we do not guarantee their fitness for your scenario and we may break compatibility in future versions.
